libsolidity/analysis/ControlFlowAnalyzer.cpp

```diff
@@ -33,131 +33,122 @@ bool ControlFlowAnalyzer::visit(FunctionDefinition const& _function)
+		auto currentNode = nodesToTraverse.top();
 		nodesToTraverse.pop();
 
-		auto& unassignedAtNode = unassigned[node];
-
-		if (node->block.returnStatement != nullptr)
-			if (node->block.returnStatement->expression())
-				unassignedAtNode.clear();
-		if (!unassignedAtNode.empty())
+		auto& nodeInfo = nodeInfos[currentNode];
+		for (auto const& variableOccurrence: currentNode->variableOccurrences)
 		{
-			// kill all return values to which a value is assigned
-			for (auto const* variableDeclaration: variablesAssignedInNode(node))
-				unassignedAtNode.erase(variableDeclaration);
-
-			// kill all return values referenced in inline assembly
-			// a reference is enough, checking whether there actually was an assignment might be overkill
-			for (auto assembly: node->block.inlineAssemblyStatements)
-				for (auto const& ref: assembly->annotation().externalReferences)
-					if (auto variableDeclaration = dynamic_cast<VariableDeclaration const*>(ref.second.declaration))
-						unassignedAtNode.erase(variableDeclaration);
+			switch (variableOccurrence.kind())
+			{
+				case VariableOccurrence::Kind::Assignment:
+					nodeInfo.unassignedVariables.erase(&variableOccurrence.declaration());
+					break;
+				case VariableOccurrence::Kind::InlineAssembly:
+					// We consider all variable referenced in inline assembly as assignments.
+					// So far any reference is enough, but we might want to check whether
+					// there actually was an assignment in the future.
+					nodeInfo.unassignedVariables.erase(&variableOccurrence.declaration());
+					break;
+				case VariableOccurrence::Kind::Access:
+					if (nodeInfo.unassignedVariables.count(&variableOccurrence.declaration()))
+					{
+						if (variableOccurrence.declaration().type()->dataStoredIn(DataLocation::Storage))
+							// Merely store the unassigned access. We do not generate an error right away, since this
+							// path might still always revert. It is only an error if this is propagated to the exit
+							// node of the function (i.e. there is a path with an uninitialized access).
+							nodeInfo.uninitializedVariableAccesses.insert(&variableOccurrence);
+					}
+					break;
+				case VariableOccurrence::Kind::Declaration:
+					nodeInfo.unassignedVariables.insert(&variableOccurrence.declaration());
+					break;
+			}
 		}
 
-		for (auto const& exit: node->exits)
-		{
-			auto& unassignedAtExit = unassigned[exit];
-			auto oldSize = unassignedAtExit.size();
-			unassignedAtExit.insert(unassignedAtNode.begin(), unassignedAtNode.end());
-			// (re)traverse an exit, if we are on a path with new unassigned return values to consider
-			// this will terminate, since there is only a finite number of unassigned return values
-			if (unassignedAtExit.size() > oldSize)
+		// Propagate changes to all exits and queue them for traversal, if needed.
+		for (auto const& exit: currentNode->exits)
+			if (nodeInfos[exit].propagateFrom(nodeInfo))
 				nodesToTraverse.push(exit);
```

Comment:
Does it make sense to turn `nodesToTraverse` to a set and only append if it is not already there?

Reply:
No - consider this:
```
if(some_condition) {} else { b = s; }
return b;
```
If I first visit return from the "else"-case everything is fine, so I also want to visit if from the "true"-case as well.
Or:
```
if(some_condition) { a = s; } else { b = s; }
return (a,b);
```
I want to arrive at the return with both ``a`` and ``b`` marked (potentially) uninitialized.

I want to walk through all nodes with the maximal set of (potentially) uninitialized variable accesses, so as long as I'm adding new variables, I have to keep traversing again.

Reply:
Ah no, I misunderstood what you meant... maybe it does make sense.